### ISSUE TITLE
[MVC/MING] Exuberant ctags  does not process if_perl.xs

### ISSUE THREAD

**Issue by ghost** [vim/vim#3526]
Exuberant ctags does not map the file extension .xs to the C language as it should and so it does not process the file if_perl.xs.  This file is referenced by the "tags" target recipe in Make_mvc.mak and Make_cyg_ming.mak.  

To correct this, the option "--langmap=c:+.xs" should be added to the CTAGS make variable definition.

**Comment by mattn** [comment]
This is not an issue of vim. please file new issue on https://github.com/universal-ctags/ctags

**Comment by mattn** [comment]
Ah, sorry. You mean that it can do it with specify option.

**Comment by k-takata** [comment]
Hm, isn't it the same for the `Makefile` for unix like systems?

P.S. It seems that you are not familiar with markdown. Use a pair of triple backquotes to surround your code.
E.g.

    ```diff
    (Paste your diff code here)
    ```

**Comment by k-takata** [comment]
Maybe, like this?

```diff
diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -106,7 +106,7 @@ endif
 
 ifndef CTAGS
 # this assumes ctags is Exuberant ctags
-CTAGS = ctags -I INIT+ --fields=+S
+CTAGS = ctags -I INIT+ --fields=+S --langmap=c:+.xs
 endif
 
 # Link against the shared version of libstdc++ by default.  Set
diff --git a/src/Make_mvc.mak b/src/Make_mvc.mak
--- a/src/Make_mvc.mak
+++ b/src/Make_mvc.mak
@@ -355,7 +355,7 @@ FEATURES = HUGE
 
 !ifndef CTAGS
 # this assumes ctags is Exuberant ctags
-CTAGS = ctags -I INIT+ --fields=+S
+CTAGS = ctags -I INIT+ --fields=+S --langmap=c:+.xs
 !endif
 
 !ifndef CSCOPE
diff --git a/src/configure.ac b/src/configure.ac
--- a/src/configure.ac
+++ b/src/configure.ac
@@ -4192,11 +4192,11 @@ dnl -i+m to test for older Exuberant cta
 AC_MSG_CHECKING(how to create tags)
 test -f tags && mv tags tags.save
 if (eval ctags --version /dev/null | grep Exuberant) < /dev/null 1>&AC_FD_CC 2>&1; then
-  TAGPRG="ctags -I INIT+ --fields=+S"
+  TAGPRG="ctags -I INIT+ --fields=+S --langmap=c:+.xs"
 elif (eval exctags --version /dev/null | grep Exuberant) < /dev/null 1>&AC_FD_CC 2>&1; then
-  TAGPRG="exctags -I INIT+ --fields=+S"
+  TAGPRG="exctags -I INIT+ --fields=+S --langmap=c:+.xs"
 elif (eval exuberant-ctags --version /dev/null | grep Exuberant) < /dev/null 1>&AC_FD_CC 2>&1; then
-  TAGPRG="exuberant-ctags -I INIT+ --fields=+S"
+  TAGPRG="exuberant-ctags -I INIT+ --fields=+S --langmap=c:+.xs"
 else
   TAGPRG="ctags"
   (eval etags	   /dev/null) < /dev/null 1>&AC_FD_CC 2>&1 && TAGPRG="etags"
```

(not tested)

**Comment by brammool** [comment]
> Exuberant ctags does not map the file extension .xs to the C language
> as it should and so it does not process the file if_perl.xs.  This
> file is referenced by the "tags" target recipe in Make_mvc.mak and
> Make_cyg_ming.mak.  
> 
> To correct this, the option "--langmap=c:+.xs" should be added to the
> CTAGS make variable definition:

That does not appear to work correctly.  E.g. it adds a tag for
do_cmdline_cmd(), while it's only called, not defined.


-- 
hundred-and-one symptoms of being an internet addict:
181. You still remember the "happy tune" your modem made
     while dialing your ISP.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by vim-ml** [comment]
On 10/10/2018 5:01 PM, Bram Moolenaar (Vim Github Repository) wrote:
>
> > Exuberant ctags does not map the file extension .xs to the C language
> > as it should and so it does not process the file if_perl.xs. This
> > file is referenced by the "tags" target recipe in Make_mvc.mak and
> > Make_cyg_ming.mak.
> >
> > To correct this, the option "--langmap=c:+.xs" should be added to the
> > CTAGS make variable definition:
>
> That does not appear to work correctly. E.g. it adds a tag for
> do_cmdline_cmd(), while it's only called, not defined.
I assumed is was C without looking too closely (truth be told, it was a 
casual look).

Apparently it is to processed by something called an xsubpp compiler?  
Regardless, it does not appear to be part of perl support in Exuberant 
ctags and changing my suggestion to --langmap=perl:+.xs does  not lead 
to a desirable result either.

**Comment by tonymec** [comment]
Maybe you should make sure to include `auto/if_perl.c` in ctags's input files instead? It is not the ultimate source, but it is human-readable and it includes here and there `#line` directives for itself and for `if_perl.xs` (the first one is at line 5 and the next one -in 57- is 61% down the file).

**Comment by vim-ml** [comment]
On 10/10/2018 9:42 PM, Tony Mechelynck (Vim Github Repository) wrote:
>
> Maybe you should make sure to include |auto/if_perl.c| in ctags's 
> input files instead? It is not the ultimate source, but it is 
> human-readable and it includes here and there |#line| directives for 
> itself and for |if_perl.xs| (the first one is at line 5 and the next 
> one -in 57- is 61% down the file).
>
I finally looked at the Makefiles for MVC and MINGW which I should have 
done before I submitted this issue.   if_perl.xs is a prerequisite for 
if_perl.c.  This says to me that if_perl.xs should be removed from the 
ctags command line because if_perl.c will be pulled in by the wildcard *.c.

**Comment by tonymec** [comment]
> On 10/10/2018 9:42 PM, Tony Mechelynck (Vim Github Repository) wrote:
> > Maybe you should make sure to include `auto/if_perl.c` in ctags's input files instead? It is not the ultimate source, but it is human-readable and it includes here and there `#line` directives for itself and for `if_perl.xs` (the first one is at line 5 and the next one -in 57- is 61% down the file).

> I finally looked at the Makefiles for MVC and MINGW which I should have done before I submitted this issue.   `if_perl.xs` is a prerequisite for `if_perl.c`.  This says to me that `if_perl.xs` should be removed from the ctags command line because `if_perl.c` will be pulled in by the wildcard `*.c`.

However, `if_perl.c` is in the `src/auto/` subdirectory (or, if you use a shadow directory, in that shadowdir's `auto/` subdirectory), unlike most of the sources which are in `src/`. In order to see `if_perl.c`, therefore, ctags must be told to look in that subdirectory. If you run ctags either in `src/` or in a shadowdir, `./*.c ./*.h ./*.pro` will pull the rest of the sources, but you need `auto/*.c` or `auto/if_perl.c` in order to pull that one.

Or at least that's how it happens on Linux where I am.

Best regards,
Tony.

**Comment by k-takata** [comment]
If we use `Make_mvc.mak` or `Make_cyg_ming.mak`, `if_perl.c` is generated under the `src` directory.
If we use `Makefile`, it is generated under the `src/auto` directory.
(It might be better always generate it under `src/auto`. Not sure if we can do it easily.)

**Comment by k-takata** [comment]
I created a PR #3540 for the difference of the directory where `if_perl.c` is generated.

**Comment by k-takata** [comment]
@brammool 

> That does not appear to work correctly. E.g. it adds a tag for do_cmdline_cmd(), while it's only called, not defined.

With or without adding the option `--langmap=c:+.xs`, a tag for `do_cmdline_cmd()` is added (which is defined in `ex_docmd.c`). I see no difference there.
Hmm, but it adds tags for `i`, `vimwin` and some other local variables which look like global variables when parsing *.xs as *.c. It is not so good, I think.

**Comment by k-takata** [comment]
I think this was fixed in 8.1.0610: https://github.com/vim/vim/commit/6dc6703295362e0d4b81a3eceae6b0dd229b5d7e.
Now we process auto/if_perl.c instead of if_perl.xs. Closing.
